### ISSUE TITLE
Finetuning only vision model

### ISSUE THREAD

**Issue by RishabhAdiga** [2U1/Qwen2-VL-Finetune#118]
Is there a way to finetune only the vision transformer and freeze the llm (no LORA or full finetuning for LLM).
And can this be done with LORA and full finetuning both (for the ViT)?

**Comment by 2U1** [comment]
Yes

1. For finetuninig only vision module, you could set the `freeze_llm` to true in the `finetune.sh`
2. Also, for the LoRA, you could add the llms layer in the lora_exclude in the `finetune_lora_vision.sh`

**Comment by strawhatboy** [comment]
@2U1 I tried `finetune_lora_vision.sh` to fine-tune Qwen2.5-VL these days, while it keeps throwing various errors which are confusing, for me, the error currently could be:
```bash
No label_names provided for model class `PeftModel`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.
```
because after this printed, the tqdm bar appeared and process ended immediately. could you help on this? many thanks!!

before this, I solved an issue that says `part of the parameters are on cpu not device[0]` by adding `model = model.to(training_args.device)` before initializing the `QwenTrainer`.
also, the `no module named "train"` issue by renaming.

**Comment by 2U1** [comment]
@strawhatboy 
`No label_names provided for model class `PeftModel`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.`
This won't be a matter it just a warning and it dosen't affect the training.

I think the module name issue is caused by the refacotring I've made. I'll take a look into it.

Thnaks for letting me know.

**Comment by strawhatboy** [comment]
> [@strawhatboy](https://github.com/strawhatboy) `No label_names provided for model class `PeftModel`. Since `PeftModel`hides base models input arguments, if label_names is not given, label_names can't be set automatically within`Trainer`. Note that empty label_names list will be used instead.` This won't be a matter it just a warning and it dosen't affect the training.
> 
> I think the module name issue is caused by the refacotring I've made. I'll take a look into it.
> 
> Thnaks for letting me know.

maybe the module name is not an issue now because I noticed that there's a commit to rename the train.py, and I solved that by renaming the whole foldername, currently for me, the issue is after 'label names' message, the yraining cannot be started, that's weird

**Comment by 2U1** [comment]
@strawhatboy  I've tested but it works fine for me. I'm not sure what is the reason. I'll test a bit more.

**Comment by strawhatboy** [comment]
> [@strawhatboy](https://github.com/strawhatboy) I've tested but it works fine for me. I'm not sure what is the reason. I'll test a bit more.

I've no idea.. here is my logs:
```bash
[2025-04-28 21:40:36,731] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to cuda (auto detect)                                                                                          [2025-04-28 21:40:39,129] [WARNING] [runner.py:215:fetch_hostfile] Unable to find hostfile, will proceed with training with local resources only.                                                                [2025-04-28 21:40:39,129] [INFO] [runner.py:607:main] cmd = /home/cuichuan/miniconda3/envs/qwen2/bin/python -u -m deepspeed.launcher.launch --world_info=eyJsb2NhbGhvc3QiOiBbMCwgMV19 --master_addr=127.0.0.1 --master_port=29500 --enable_each_rank_log=None src/train/train_sft.py --use_liger True --lora_enable True --vision_lora True --use_dora False --lora_namespan_exclude ['lm_head', 'embed_tokens'] --lora_rank 64 --lora_alpha 64 --lora_dropout 0.05 --num_lora_modules -1 --deepspeed scripts/zero2_offload.json --model_id Qwen/Qwen2.5-VL-3B-Instruct --data_path /home/cuichuan/data/agiqa-3k/directly_train_cls_minicpm_quality_multiround-llava.jsonl --image_folder /home/cuichuan/data/agiqa-3k/images --remove_unused_columns False --freeze_vision_tower True --freeze_llm True --freeze_merger True --bf16 True --fp16 False --disable_flash_attn2 False --output_dir output/lora_vision_test --num_train_epochs 1 --per_device_train_batch_size 2 --gradient_accumulation_steps 1 --image_min_pixels 200704 --image_max_pixels 1003520 --learning_rate 2e-4 --weight_decay 0.1 --warmup_ratio 0.03 --lr_scheduler_type cosine --logging_steps 1 --tf32 True --gradient_checkpointing True --report_to tensorboard --lazy_preprocess True --save_strategy steps --save_steps 200 --save_total_limit 10 --dataloader_num_workers 4                                                                                                                                                            [2025-04-28 21:40:40,514] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to cuda (auto detect)                                                                                          [2025-04-28 21:40:42,917] [INFO] [launch.py:146:main] WORLD INFO DICT: {'localhost': [0, 1]}                                                                                                                     [2025-04-28 21:40:42,917] [INFO] [launch.py:152:main] nnodes=1, num_local_procs=2, node_rank=0                                                                                                                   [2025-04-28 21:40:42,917] [INFO] [launch.py:163:main] global_rank_mapping=defaultdict(<class 'list'>, {'localhost': [0, 1]})                                                                                     [2025-04-28 21:40:42,917] [INFO] [launch.py:164:main] dist_world_size=2                                                                                                                                          [2025-04-28 21:40:42,917] [INFO] [launch.py:168:main] Setting CUDA_VISIBLE_DEVICES=0,1                                                                                                                           [2025-04-28 21:40:42,918] [INFO] [launch.py:256:main] process 2415499 spawned with command: ['/home/cuichuan/miniconda3/envs/qwen2/bin/python', '-u', 'src/train/train_sft.py', '--local_rank=0', '--use_liger', 'True', '--lora_enable', 'True', '--vision_lora', 'True', '--use_dora', 'False', '--lora_namespan_exclude', "['lm_head', 'embed_tokens']", '--lora_rank', '64', '--lora_alpha', '64', '--lora_dropout', '0.05', '--num_lora_modules', '-1', '--deepspeed', 'scripts/zero2_offload.json', '--model_id', 'Qwen/Qwen2.5-VL-3B-Instruct', '--data_path', '/home/cuichuan/data/agiqa-3k/directly_train_cls_minicpm_quality_multiround-llava.jsonl', '--image_folder', '/home/cuichuan/data/agiqa-3k/images', '--remove_unused_columns', 'False', '--freeze_vision_tower', 'True', '--freeze_llm', 'True', '--freeze_merger', 'True', '--bf16', 'True', '--fp16', 'False', '--disable_flash_attn2', 'False', '--output_dir', 'output/lora_vision_test', '--num_train_epochs', '1', '--per_device_train_batch_size', '2', '--gradient_accumulation_steps', '1', '--image_min_pixels', '200704', '--image_max_pixels', '1003520', '--learning_rate', '2e-4', '--weight_decay', '0.1', '--warmup_ratio', '0.03', '--lr_scheduler_type', 'cosine', '--logging_steps', '1', '--tf32', 'True', '--gradient_checkpointing', 'True', '--report_to', 'tensorboard', '--lazy_preprocess', 'True', '--save_strategy', 'steps', '--save_steps', '200', '--save_total_limit', '10', '--dataloader_num_workers', '4']     [2025-04-28 21:40:42,919] [INFO] [launch.py:256:main] process 2415500 spawned with command: ['/home/cuichuan/miniconda3/envs/qwen2/bin/python', '-u', 'src/train/train_sft.py', '--local_rank=1', '--use_liger', 'True', '--lora_enable', 'True', '--vision_lora', 'True', '--use_dora', 'False', '--lora_namespan_exclude', "['lm_head', 'embed_tokens']", '--lora_rank', '64', '--lora_alpha', '64', '--lora_dropout', '0.05', '--num_lora_modules', '-1', '--deepspeed', 'scripts/zero2_offload.json', '--model_id', 'Qwen/Qwen2.5-VL-3B-Instruct', '--data_path', '/home/cuichuan/data/agiqa-3k/directly_train_cls_minicpm_quality_multiround-llava.jsonl', '--image_folder', '/home/cuichuan/data/agiqa-3k/images', '--remove_unused_columns', 'False', '--freeze_vision_tower', 'True', '--freeze_llm', 'True', '--freeze_merger', 'True', '--bf16', 'True', '--fp16', 'False', '--disable_flash_attn2', 'False', '--output_dir', 'output/lora_vision_test', '--num_train_epochs', '1', '--per_device_train_batch_size', '2', '--gradient_accumulation_steps', '1', '--image_min_pixels', '200704', '--image_max_pixels', '1003520', '--learning_rate', '2e-4', '--weight_decay', '0.1', '--warmup_ratio', '0.03', '--lr_scheduler_type', 'cosine', '--logging_steps', '1', '--tf32', 'True', '--gradient_checkpointing', 'True', '--report_to', 'tensorboard', '--lazy_preprocess', 'True', '--save_strategy', 'steps', '--save_steps', '200', '--save_total_limit', '10', '--dataloader_num_workers', '4']     [2025-04-28 21:40:46,732] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to cuda (auto detect)                                                                                          [2025-04-28 21:40:46,745] [INFO] [real_accelerator.py:222:get_accelerator] Setting ds_accelerator to cuda (auto detect)                                                                                          [2025-04-28 21:40:47,710] [INFO] [comm.py:658:init_distributed] cdb=None                                                                                                                                         [2025-04-28 21:40:47,711] [INFO] [comm.py:689:init_distributed] Initializing TorchBackend in DeepSpeed with backend nccl
[2025-04-28 21:40:47,710] [INFO] [comm.py:658:init_distributed] cdb=None                                                                                                                                [52/1972][2025-04-28 21:40:47,711] [INFO] [comm.py:689:init_distributed] Initializing TorchBackend in DeepSpeed with backend nccl                                                                                         [2025-04-28 21:40:47,821] [INFO] [comm.py:658:init_distributed] cdb=None                                                                                                                                         You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.                                      Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  4.22it/s]You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.                                      Loading checkpoint shards: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 2/2 [00:00<00:00,  4.70it/s]Found 414 lora modules: ['visual.blocks.0.attn.qkv', 'visual.blocks.0.attn.proj', 'visual.blocks.0.mlp.gate_proj', 'visual.blocks.0.mlp.up_proj', 'visual.blocks.0.mlp.down_proj', 'visual.blocks.1.attn.qkv', 'visual.blocks.1.attn.proj', 'visual.blocks.1.mlp.gate_proj',...
...
del.layers.35.self_attn.o_proj', 'model.layers.35.mlp.gate_proj', 'model.layers.35.mlp.up_proj', 'model.layers.35.mlp.down_proj']                                                                                Adding LoRA to the model...                                                                                                                                                                                      Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.50, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.                                                                                               Using a slow image processor as `use_fast` is unset and a slow processor was saved with this model. `use_fast=True` will be the default behavior in v4.50, even if the model was saved with a slow processor. This will result in minor differences in outputs. You'll still be able to use a slow processor with `use_fast=False`.                                                                                               No label_names provided for model class `PeftModel`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.                                                                                                                                                                            No label_names provided for model class `PeftModel`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.                                                                                                                                                                            Using /home/cuichuan/.cache/torch_extensions/py310_cu124 as PyTorch extensions root...                                                                                                                           Emitting ninja build file /home/cuichuan/.cache/torch_extensions/py310_cu124/cpu_adam/build.ninja...                                                                                                             Building extension module cpu_adam...                                                                                                                                                                            Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)                                                                                                ninja: no work to do.                                                                                                                                                                                            Loading extension module cpu_adam...                                                                                                                                                                             Time to load cpu_adam op: 2.3195576667785645 seconds                                                                                                                                                             Using /home/cuichuan/.cache/torch_extensions/py310_cu124 as PyTorch extensions root...                                                                                                                           Emitting ninja build file /home/cuichuan/.cache/torch_extensions/py310_cu124/cpu_adam/build.ninja...                                                                                                             Building extension module cpu_adam...                                                                                                                                                                            Allowing ninja to set a default number of workers... (overridable by setting the environment variable MAX_JOBS=N)                                                                                                ninja: no work to do.                                                                                                                                                                                            Loading extension module cpu_adam...                                                                                                                                                                             Time to load cpu_adam op: 2.3169658184051514 seconds                                                                                                                                                             [2025-04-28 21:41:10,948] [INFO] [launch.py:319:sigkill_handler] Killing subprocess 2415499                                                                                                                      [2025-04-28 21:41:10,949] [INFO] [launch.py:319:sigkill_handler] Killing subprocess 2415500                                                                                                                      [2025-04-28 21:41:10,963] [ERROR] [launch.py:325:sigkill_handler] ['/home/cuichuan/miniconda3/envs/qwen2/bin/python', '-u', 'src/train/train_sft.py', '--local_rank=1', '--use_liger', 'True', '--lora_enable', 'True', '--vision_lora', 'True', '--use_dora', 'False', '--lora_namespan_exclude', "['lm_head', 'embed_tokens']", '--lora_rank', '64', '--lora_alpha', '64', '--lora_dropout', '0.05', '--num_lora_modules', '-1', '--deepspeed', 'scripts/zero2_offload.json', '--model_id', 'Qwen/Qwen2.5-VL-3B-Instruct', '--data_path', '/home/cuichuan/data/agiqa-3k/directly_train_cls_minicpm_quality_multiround-llava.jsonl', '--image_folder', '/home/cuichuan/data/agiqa-3k/images', '--remove_unused_columns', 'False', '--freeze_vision_tower', 'True', '--freeze_llm', 'True', '--freeze_merger', 'True', '--bf16', 'True', '--fp16', 'False', '--disable_flash_attn2', 'False', '--output_dir', 'output/lora_vision_test', '--num_train_epochs', '1', '--per_device_train_batch_size', '2', '--gradient_accumulation_steps', '1', '--image_min_pixels', '200704', '--image_max_pixels', '1003520', '--learning_rate', '2e-4', '--weight_decay', '0.1', '--warmup_ratio', '0.03', '--lr_scheduler_type', 'cosine', '--logging_steps', '1', '--tf32', 'True', '--gradient_checkpointing', 'True', '--report_to', 'tensorboard', '--lazy_preprocess', 'True', '--save_strategy', 'steps', '--save_steps', '200', '--save_total_limit', '10', '--dataloader_num_workers', '4'] exits with return code = -11
```

it returns -11 error code, after adding line `model = model.to(training_args.device)`, it returns -9 instead. 😟

**Comment by 2U1** [comment]
-9 usually occurs when system memory oom.
I think erasing the `model = model.to(training_args.device)` would be better.
